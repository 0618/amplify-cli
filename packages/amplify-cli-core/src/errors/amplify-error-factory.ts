/**
 * This class is a placeholder for filtering errors coming from @aws-amplify/amplify-category-api package.
 * This whole class can be removed if the above package implements amplify Errors and AmplifyFaults
 */
import { $TSAny } from '../types';
import { AmplifyError } from './amplify-error';
import { AmplifyErrorType } from './amplify-exception';
import { AmplifyFault } from './amplify-fault';

/**
 * Error types defined in @aws-amplify/amplify-category-api
 */
const amplifyErrorList = [
  'InvalidDirectiveError' ,
  'InvalidTransformerError',
  'SchemaValidationError',
  'TransformerContractError',
  'DestructiveMigrationError',
  'InvalidMigrationError',
  'InvalidGSIMigrationError', 
  'UnknownDirectiveError', 
  'GraphQLError'
 ];
 
 const schemaNotFoundRegexError = [
     /^.*\bNo GraphQL schema found\b.*$/,
     /^.*\bNo schema found\b.*$/,
     /^.*\bCould not find a schema\b.*$/
 ];
 
 const apiAlreadyExistsRegexError = [
   /^.*\bUse 'amplify update api' to make modifications\b.*$/,
 ];
 
// error can be an AmplifyException Type or  Error type or any type
export class AmplifyErrorFactory {
    private readonly defaultError: AmplifyFault;
    constructor(error: AmplifyFault) {
        this.defaultError = error;
    }
    create(error: $TSAny): AmplifyError | AmplifyFault {
      let amplifyError;
        if(error instanceof Error && error?.name && amplifyErrorList.includes(error.name)){
            const amplifyErrorType = `${error.name}` as AmplifyErrorType;
            amplifyError =  new AmplifyError(
                amplifyErrorType,
                {
                  message: error.message,
                },
                error,
              )
        }
        // try regex matching for errors without name 
        const regexUserArrayMap = this.createMapFromRegexList();
        for (const [regexArray, errorName] of regexUserArrayMap.entries()) {
          regexArray.forEach( regex => {
            if(regex.test(error.message)){
              amplifyError =  new AmplifyError(
                errorName as AmplifyErrorType,
                {
                  message: error.message,
                },
                error,
              ) 
            }
          });
        }
        if(amplifyError){
          return amplifyError;
        }
        return this.defaultError;
    }

    /**
     * creates a map of Array<Regexp> and Error key to group a section of erros to a specific bucket
     */
    private createMapFromRegexList(): Map<Array<RegExp>,string> {
      const regexUserArrayMap = new Map<Array<RegExp>,string>();
      // setting SchemaNotFoundError here
      regexUserArrayMap.set(schemaNotFoundRegexError,'SchemaNotFoundError');
      regexUserArrayMap.set(apiAlreadyExistsRegexError, 'ResourceAlreadyExistsError')
      return regexUserArrayMap;
    }
}


