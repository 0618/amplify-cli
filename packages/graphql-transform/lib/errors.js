"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var InvalidTransformerError = /** @class */ (function (_super) {
    __extends(InvalidTransformerError, _super);
    function InvalidTransformerError(message) {
        var _this = _super.call(this, message) || this;
        _this.name = "InvalidTransformerError";
        if (Error.captureStackTrace) {
            Error.captureStackTrace(_this, InvalidTransformerError);
        }
        return _this;
    }
    return InvalidTransformerError;
}(Error));
exports.InvalidTransformerError = InvalidTransformerError;
var InvalidDirectiveDefinitionError = /** @class */ (function (_super) {
    __extends(InvalidDirectiveDefinitionError, _super);
    function InvalidDirectiveDefinitionError(message) {
        var _this = _super.call(this, message) || this;
        _this.name = "InvalidDirectiveDefinitionError";
        if (Error.captureStackTrace) {
            Error.captureStackTrace(_this, InvalidDirectiveDefinitionError);
        }
        return _this;
    }
    return InvalidDirectiveDefinitionError;
}(Error));
exports.InvalidDirectiveDefinitionError = InvalidDirectiveDefinitionError;
var InvalidDirectiveError = /** @class */ (function (_super) {
    __extends(InvalidDirectiveError, _super);
    function InvalidDirectiveError(message) {
        var _this = _super.call(this, message) || this;
        _this.name = "InvalidDirectiveError";
        if (Error.captureStackTrace) {
            Error.captureStackTrace(_this, InvalidDirectiveError);
        }
        return _this;
    }
    return InvalidDirectiveError;
}(Error));
exports.InvalidDirectiveError = InvalidDirectiveError;
var UnknownDirectiveError = /** @class */ (function (_super) {
    __extends(UnknownDirectiveError, _super);
    function UnknownDirectiveError(message) {
        var _this = _super.call(this, message) || this;
        _this.name = "UnknownDirectiveError";
        if (Error.captureStackTrace) {
            Error.captureStackTrace(_this, UnknownDirectiveError);
        }
        return _this;
    }
    return UnknownDirectiveError;
}(Error));
exports.UnknownDirectiveError = UnknownDirectiveError;
//# sourceMappingURL=errors.js.map