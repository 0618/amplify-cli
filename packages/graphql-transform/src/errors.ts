import { GraphQLError } from 'graphql'

export class InvalidTransformerError extends Error {

    constructor(message: string) {
        super(message);
        this.name = "InvalidTransformerError";
        if ((Error as any).captureStackTrace) {
            (Error as any).captureStackTrace(this, InvalidTransformerError)
        }
    }
}

export class TransformSchemaError extends Error {

    constructor(errors: GraphQLError[]) {
        super(`Schema Errors:\n\n${errors.join('\n')}`);
        this.name = "TransformSchemaError";
        if ((Error as any).captureStackTrace) {
            (Error as any).captureStackTrace(this, TransformSchemaError)
        }
    }
}

export class InvalidDirectiveError extends Error {
    constructor(message: string) {
        super(message);
        this.name = "InvalidDirectiveError";
        if ((Error as any).captureStackTrace) {
            (Error as any).captureStackTrace(this, InvalidDirectiveError)
        }
    }
}

export class UnknownDirectiveError extends Error {
    constructor(message: string) {
        super(message);
        this.name = "UnknownDirectiveError";
        if ((Error as any).captureStackTrace) {
            (Error as any).captureStackTrace(this, UnknownDirectiveError)
        }
    }
}
