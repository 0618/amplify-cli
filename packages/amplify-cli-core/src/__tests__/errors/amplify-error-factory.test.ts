import { AmplifyError } from '../../errors/amplify-error';
import { AmplifyErrorFactory } from '../../errors/amplify-error-factory';
import { AmplifyErrorType, AmplifyException } from '../../errors/amplify-exception';
import { AmplifyFault } from '../../errors/amplify-fault';


const errorType: AmplifyErrorType = 'DeploymentError';
// converted error to amplifyException
const error = new Error('mockMessage');
error.name = errorType;
/**
 * deepest amplify exception to get printed if amplify fault is thrown
 */
const deepestAmplifyException = new AmplifyError(
    errorType,
  {
    message: error.message,
  },
  error,
);

const amplifyFault = new AmplifyFault(
  'DeploymentFault',
  {
    message: 'mockMessage',
  },
  deepestAmplifyException,
);

const amplifyError = new AmplifyError(
    'AmplifyStudioError',
    {
      message: 'mockMessage',
    },
  );

test('returns amplify fault if the error isnt mentioned in list', async () => {
  expect(getDeepestAmplifyException(new AmplifyErrorFactory(amplifyFault).create(error)).name).toMatch(errorType);
});

test('returns amplify fault if the error isnt instance of Error', async () => {
  expect(getDeepestAmplifyException(new AmplifyErrorFactory(amplifyFault).create(amplifyError)).name).toMatch(errorType);
});

test('returns user error if the error is present in list', async () => {
    // error name is user error list
  error.name = 'InvalidDirectiveError';
  expect(getDeepestAmplifyException(new AmplifyErrorFactory(amplifyFault).create(error)).name).toMatch('InvalidDirectiveError')
});

test('returns regex error if the error contains regex user mapping', async () => {
    const testUserErrorString = [
        'No GraphQL schema found in schema location',
        'No schema found, your graphql schema should be in either',
        'Could not find a schema in either',
        'Api Not found',
        `GraphQL API myMockAppsyncApi already exists in the project. Use 'amplify update api' to make modifications.`,
        `GraphQL API mockapi2 already exists.`
    ];
    // making name as null
    error.name = '';
    // error name is user error list
    testUserErrorString.forEach(testString => {
        error.message = testString;
        expect(getDeepestAmplifyException(new AmplifyErrorFactory(amplifyFault).create(error)).name).toMatchSnapshot()
    });
});

const getDeepestAmplifyException = (amplifyException: AmplifyException): AmplifyException => {
  let deepestAmplifyException = amplifyException;
  while (deepestAmplifyException.downstreamException && deepestAmplifyException.downstreamException instanceof AmplifyException) {
    deepestAmplifyException = deepestAmplifyException.downstreamException;
  }
  return deepestAmplifyException;
};
