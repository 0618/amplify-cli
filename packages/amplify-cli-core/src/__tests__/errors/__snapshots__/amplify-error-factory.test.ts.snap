// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns regex error if the error contains regex user mapping 1`] = `"SchemaNotFoundError"`;

exports[`returns regex error if the error contains regex user mapping 2`] = `"SchemaNotFoundError"`;

exports[`returns regex error if the error contains regex user mapping 3`] = `"SchemaNotFoundError"`;

exports[`returns regex error if the error contains regex user mapping 4`] = `"DeploymentError"`;

exports[`returns regex error if the error contains regex user mapping 5`] = `"ResourceAlreadyExistsError"`;

exports[`returns regex error if the error contains regex user mapping 6`] = `"DeploymentError"`;
